<a href="https://colab.research.google.com/github/shubhanalytics/PGPDSBA-Projects/blob/main/Proj.01%20-%20Python%20for%20Data%20Analysis/Uber%20Drive.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

UBER DRIVE ANALYSIS : The project is based on the trips made by Uber drivers. Different aspects of the trip are analyzed by using different functions in Python.

Load the necessary libraries. Import and load the dataset with a name uber.

In [1]:
import numpy as np, pandas as pd, seaborn as sns, os
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
print(os.getcwd()) # checking the current working directory and printing in the output to get confirmed

D:\My Data\Professional\Analytics\PGPDSBA-Projects\Proj.01 - Python for Data Analysis


In [3]:
# load the csv file uberdrive.csv
uber = pd.read_csv(r'D:\My Data\Professional\Analytics\PGPDSBA-Projects\Proj.01 - Python for Data Analysis\Uber_Drive.csv') 

uber.head() #read the file and check the first 5 rows

,START_DATE*,END_DATE*,CATEGORY*,START*,STOP*,MILES*,PURPOSE*
0,1/1/2016 21:11,1/1/2016 21:17,Business,Fort Pierce,Fort Pierce,5.1,Meal/Entertain
1,1/2/2016 1:25,1/2/2016 1:37,Business,Fort Pierce,Fort Pierce,5.0,NaN
2,1/2/2016 20:25,1/2/2016 20:38,Business,Fort Pierce,Fort Pierce,4.8,Errand/Supplies
3,1/5/2016 17:31,1/5/2016 17:45,Business,Fort Pierce,Fort Pierce,4.7,Meeting
4,1/6/2016 14:42,1/6/2016 15:49,Business,Fort Pierce,West Palm Beach,63.7,Customer Visit


In [4]:
uber.tail(10)   # Displaying the last 10 records of the dataset

,START_DATE*,END_DATE*,CATEGORY*,START*,STOP*,MILES*,PURPOSE*
1145,12/30/2016 10:15,12/30/2016 10:33,Business,Karachi,Karachi,2.8,Errand/Supplies
1146,12/30/2016 11:31,12/30/2016 11:56,Business,Karachi,Karachi,2.9,Errand/Supplies
1147,12/30/2016 15:41,12/30/2016 16:03,Business,Karachi,Karachi,4.6,Errand/Supplies
1148,12/30/2016 16:45,12/30/2016 17:08,Business,Karachi,Karachi,4.6,Meeting
1149,12/30/2016 23:06,12/30/2016 23:10,Business,Karachi,Karachi,0.8,Customer Visit
1150,12/31/2016 1:07,12/31/2016 1:14,Business,Karachi,Karachi,0.7,Meeting
1151,12/31/2016 13:24,12/31/2016 13:42,Business,Karachi,Unknown Location,3.9,Temporary Site
1152,12/31/2016 15:03,12/31/2016 15:38,Business,Unknown Location,Unknown Location,16.2,Meeting
1153,12/31/2016 21:32,12/31/2016 21:50,Business,Katunayake,Gampaha,6.4,Temporary Site
1154,12/31/2016 22:08,12/31/2016 23:51,Business,Gampaha,Ilukwatta,48.2,Temporary Site


In [5]:
uber.shape  # showing the shape or dimension(number of rows, number of columns) of the dataset

(1155, 7)

In [6]:
uber.size   # showing the size(rows * columns) of the dataset

8085

In [7]:
uber.info()    # displaying the information about all the variables of the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1155 entries, 0 to 1154
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   START_DATE*  1155 non-null   object 
 1   END_DATE*    1155 non-null   object 
 2   CATEGORY*    1155 non-null   object 
 3   START*       1155 non-null   object 
 4   STOP*        1155 non-null   object 
 5   MILES*       1155 non-null   float64
 6   PURPOSE*     653 non-null    object 
dtypes: float64(1), object(6)
memory usage: 63.3+ KB


In the above dataset info, below are the few observations:

1. There are 1155 non-null values in first 6 columns, but there are only 653 non-null values in the 'purpose' column, which means 502 values are null in 'purpose' column. We need to review and correct it.

2. Only the 'Miles' column is in float datatype, rest all columns are in object datatype, which requires our manual review and correction.


In [8]:
uber.isnull().any()    # checking the missing values in the dataset in the boolean(True or False) format

START_DATE*    False
END_DATE*      False
CATEGORY*      False
START*         False
STOP*          False
MILES*         False
PURPOSE*        True
dtype: bool

In [9]:
uber.isnull().sum()  # checking the missing values in the dataset in the integer format

START_DATE*      0
END_DATE*        0
CATEGORY*        0
START*           0
STOP*            0
MILES*           0
PURPOSE*       502
dtype: int64

In [10]:
uber.isnull().sum().sum()   # displaying the total number of missing values across all the variables

np.int64(502)

In [11]:
uber.describe()   # Summary of the original data including : Count,Mean, Std, Min, 25%,50%,75% and max

,MILES*
count,1155.000000
mean,10.566840
std,21.579106
min,0.500000
25%,2.900000
50%,6.000000
75%,10.400000
max,310.300000


In the above summary, we can see that the maximum miles driven by an uber drive is 310, and the minimum miles is 0.50

In [12]:
df = uber.copy()  # copied the original dataset(uber) to df

In [13]:
df.head()

,START_DATE*,END_DATE*,CATEGORY*,START*,STOP*,MILES*,PURPOSE*
0,1/1/2016 21:11,1/1/2016 21:17,Business,Fort Pierce,Fort Pierce,5.1,Meal/Entertain
1,1/2/2016 1:25,1/2/2016 1:37,Business,Fort Pierce,Fort Pierce,5.0,NaN
2,1/2/2016 20:25,1/2/2016 20:38,Business,Fort Pierce,Fort Pierce,4.8,Errand/Supplies
3,1/5/2016 17:31,1/5/2016 17:45,Business,Fort Pierce,Fort Pierce,4.7,Meeting
4,1/6/2016 14:42,1/6/2016 15:49,Business,Fort Pierce,West Palm Beach,63.7,Customer Visit


In [14]:
df = df.dropna() # dropped the missing values from the df

df.head()  #display 'df' which does not contain any missing values

,START_DATE*,END_DATE*,CATEGORY*,START*,STOP*,MILES*,PURPOSE*
0,1/1/2016 21:11,1/1/2016 21:17,Business,Fort Pierce,Fort Pierce,5.1,Meal/Entertain
2,1/2/2016 20:25,1/2/2016 20:38,Business,Fort Pierce,Fort Pierce,4.8,Errand/Supplies
3,1/5/2016 17:31,1/5/2016 17:45,Business,Fort Pierce,Fort Pierce,4.7,Meeting
4,1/6/2016 14:42,1/6/2016 15:49,Business,Fort Pierce,West Palm Beach,63.7,Customer Visit
5,1/6/2016 17:15,1/6/2016 17:19,Business,West Palm Beach,West Palm Beach,4.3,Meal/Entertain


In [15]:
print(f"shape of dataset after dropped missing values: {df.shape}")
print(f"size of dataset after dropped missing values: {df.size}")

shape of dataset after dropped missing values: (653, 7)
size of dataset after dropped missing values: 4571


In [16]:
df.info()   #dataframe 'df' information

<class 'pandas.core.frame.DataFrame'>
Index: 653 entries, 0 to 1154
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   START_DATE*  653 non-null    object 
 1   END_DATE*    653 non-null    object 
 2   CATEGORY*    653 non-null    object 
 3   START*       653 non-null    object 
 4   STOP*        653 non-null    object 
 5   MILES*       653 non-null    float64
 6   PURPOSE*     653 non-null    object 
dtypes: float64(1), object(6)
memory usage: 40.8+ KB


In [17]:
new_column_names = {'START_DATE*':'start_date',\
                     'END_DATE*':'end_date',\
                     'CATEGORY*':'category',\
                     'START*':'start',\
                     'STOP*':'stop',\
                     'MILES*':'miles',\
                     'PURPOSE*':'purpose'}

In [18]:
df = df.rename(columns = new_column_names)

In [19]:
df.columns

Index(['start_date', 'end_date', 'category', 'start', 'stop', 'miles',
       'purpose'],
      dtype='object')

In [20]:
df.head()

,start_date,end_date,category,start,stop,miles,purpose
0,1/1/2016 21:11,1/1/2016 21:17,Business,Fort Pierce,Fort Pierce,5.1,Meal/Entertain
2,1/2/2016 20:25,1/2/2016 20:38,Business,Fort Pierce,Fort Pierce,4.8,Errand/Supplies
3,1/5/2016 17:31,1/5/2016 17:45,Business,Fort Pierce,Fort Pierce,4.7,Meeting
4,1/6/2016 14:42,1/6/2016 15:49,Business,Fort Pierce,West Palm Beach,63.7,Customer Visit
5,1/6/2016 17:15,1/6/2016 17:19,Business,West Palm Beach,West Palm Beach,4.3,Meal/Entertain


In [21]:
print(df['start'].nunique())   # getting the unique START places names

131


In [22]:
df['start'].unique()

array(['Fort Pierce', 'West Palm Beach', 'Cary', 'Jamaica', 'New York',
       'Elmhurst', 'Midtown', 'East Harlem', 'Flatiron District',
       'Midtown East', 'Hudson Square', 'Lower Manhattan',
       "Hell's Kitchen", 'Downtown', 'Gulfton', 'Houston', 'Eagan Park',
       'Morrisville', 'Durham', 'Farmington Woods', 'Lake Wellingborough',
       'Fayetteville Street', 'Raleigh', 'Whitebridge', 'Hazelwood',
       'Fairmont', 'Meredith Townes', 'Apex', 'Chapel Hill', 'Northwoods',
       'Edgehill Farms', 'Eastgate', 'East Elmhurst', 'Long Island City',
       'Katunayaka', 'Colombo', 'Nugegoda', 'Unknown Location',
       'Islamabad', 'R?walpindi', 'Noorpur Shahan', 'Preston',
       'Heritage Pines', 'Tanglewood', 'Waverly Place', 'Wayne Ridge',
       'Westpark Place', 'East Austin', 'The Drag', 'South Congress',
       'Georgian Acres', 'North Austin', 'West University', 'Austin',
       'Katy', 'Sharpstown', 'Sugar Land', 'Galveston', 'Port Bolivar',
       'Washington Avenue',

In [23]:
df['start'].value_counts().head(20)

start
Cary                161
Unknown Location     55
Morrisville          54
Whitebridge          36
Durham               30
Karachi              26
Raleigh              21
Lahore               19
Islamabad            15
Midtown              11
Apex                 11
Berkeley              8
San Francisco         7
The Drag              6
Westpark Place        5
Kissimmee             5
Rawalpindi            5
Orlando               4
Fort Pierce           4
Hazelwood             4
Name: count, dtype: int64

In [24]:
# displaying total number of unique START destinations of original dataframe 'uber'

uber['START*'].nunique()

176

In [25]:
# displaying total number of unique STOP destinations of original dataframe 'uber'

uber['STOP*'].nunique()

187

In [26]:
#Let's check the rides with the starting point as 'san fransisco'

# Now, selecting the index column as 'START*'

df_sf = pd.read_csv(r'D:\My Data\Professional\Analytics\PGPDSBA-Projects\Proj.01 - Python for Data Analysis\Uber_Drive.csv', index_col='START*')

# selecting the rows which has starting point of San Francisco using loc function and assign it to new variable
uber_trips_SF = df_sf.loc['San Francisco']


# displaying all the uber trips that has starting point of San Francisco
uber_trips_SF

,START_DATE*,END_DATE*,CATEGORY*,STOP*,MILES*,PURPOSE*
START*,,,,,,
San Francisco,5/9/2016 14:39,5/9/2016 15:06,Business,Palo Alto,20.5,Between Offices
San Francisco,6/14/2016 16:09,6/14/2016 16:39,Business,Emeryville,11.6,Meeting
San Francisco,10/19/2016 14:02,10/19/2016 14:31,Business,Berkeley,10.8,NaN
San Francisco,11/7/2016 19:17,11/7/2016 19:57,Business,Berkeley,13.2,Between Offices
San Francisco,11/8/2016 12:16,11/8/2016 12:49,Business,Berkeley,11.3,Meeting
San Francisco,11/9/2016 18:40,11/9/2016 19:17,Business,Oakland,12.7,Customer Visit
San Francisco,11/10/2016 15:17,11/10/2016 15:22,Business,Oakland,9.9,Temporary Site
San Francisco,11/15/2016 20:44,11/15/2016 21:00,Business,Berkeley,11.8,Temporary Site


In [27]:
uber_trips_SF.shape

(8, 6)

In [28]:
uber_trips_SF.value_counts()   # summing up all the numerical columns of the dataframe 'uber_trips_SF'

START_DATE*       END_DATE*         CATEGORY*  STOP*       MILES*  PURPOSE*       
11/10/2016 15:17  11/10/2016 15:22  Business   Oakland     9.9     Temporary Site     1
11/15/2016 20:44  11/15/2016 21:00  Business   Berkeley    11.8    Temporary Site     1
11/7/2016 19:17   11/7/2016 19:57   Business   Berkeley    13.2    Between Offices    1
11/8/2016 12:16   11/8/2016 12:49   Business   Berkeley    11.3    Meeting            1
11/9/2016 18:40   11/9/2016 19:17   Business   Oakland     12.7    Customer Visit     1
5/9/2016 14:39    5/9/2016 15:06    Business   Palo Alto   20.5    Between Offices    1
6/14/2016 16:09   6/14/2016 16:39   Business   Emeryville  11.6    Meeting            1
Name: count, dtype: int64

In [33]:
# Most popular starting point in the original dataset 'uber'

print('The most popular starting point for the uber trips is:', uber['START*'].value_counts().idxmax())
print('_____________________')
print('See the below data to verify:\n\n', uber['START*'].value_counts().sort_values(ascending=False).head(5))


The most popular starting point for the uber trips is: Cary
_____________________
See the below data to verify:

 START*
Cary                201
Unknown Location    148
Morrisville          85
Whitebridge          68
Islamabad            57
Name: count, dtype: int64


In [35]:
# most popular dropping point from the original dataset 'uber'

print('The most popular dropping point for the Uber trips is:', uber['STOP*'].value_counts().idxmax())
print('_________________________')
print('See the below data to verify:\n\n', uber['STOP*'].value_counts().sort_values(ascending=False).head(5))


The most popular dropping point for the Uber trips is: Cary
_________________________
See the below data to verify:

 STOP*
Cary                203
Unknown Location    149
Morrisville          84
Whitebridge          65
Islamabad            58
Name: count, dtype: int64


In [53]:
# most frequent route taken by uber drivers (using original dataset 'uber')

most_frequent = (uber.groupby(['START*', 'STOP*']).size().reset_index(name='count_of_trips').sort_values('count_of_trips', ascending=False).head())

most_frequent

,START*,STOP*,count_of_trips
318,Unknown Location,Unknown Location,86
203,Morrisville,Cary,75
46,Cary,Morrisville,67
40,Cary,Cary,53
42,Cary,Durham,36


In [ ]:
# pip install pandasql

In [47]:
df.query("miles > 50")

,start_date,end_date,category,start,stop,miles,purpose
4,1/6/2016 14:42,1/6/2016 15:49,Business,Fort Pierce,West Palm Beach,63.7,Customer Visit
232,3/17/2016 12:52,3/17/2016 15:11,Business,Austin,Katy,136.0,Customer Visit
251,3/19/2016 19:33,3/19/2016 20:39,Business,Galveston,Houston,57.0,Customer Visit
268,3/25/2016 13:24,3/25/2016 16:22,Business,Cary,Latta,144.0,Customer Visit
269,3/25/2016 16:52,3/25/2016 22:22,Business,Latta,Jacksonville,310.3,Customer Visit
270,3/25/2016 22:54,3/26/2016 1:39,Business,Jacksonville,Kissimmee,201.0,Meeting
295,4/2/2016 12:21,4/2/2016 14:47,Business,Kissimmee,Daytona Beach,77.3,Customer Visit
296,4/2/2016 16:57,4/2/2016 18:09,Business,Daytona Beach,Jacksonville,80.5,Customer Visit
297,4/2/2016 19:38,4/2/2016 22:36,Business,Jacksonville,Ridgeland,174.2,Customer Visit
298,4/2/2016 23:11,4/3/2016 1:34,Business,Ridgeland,Florence,144.0,Meeting


In [55]:
#displaying all types of purposes for the trip in an array from the dataframe 'df' with no 'NA' values

df['purpose'].unique()


array(['Meal/Entertain', 'Errand/Supplies', 'Meeting', 'Customer Visit',
       'Temporary Site', 'Between Offices', 'Charity ($)', 'Commute',
       'Moving', 'Airport/Travel'], dtype=object)

In [64]:
uber['PURPOSE*'].unique()

array(['Meal/Entertain', nan, 'Errand/Supplies', 'Meeting',
       'Customer Visit', 'Temporary Site', 'Between Offices',
       'Charity ($)', 'Commute', 'Moving', 'Airport/Travel'], dtype=object)

## Q19. Plot a bar graph of Purpose vs Miles(Distance). (3 points)
#### Note: Use the original dataframe without dropping the 'NA' values.
#### Hint:You have to plot total/sum miles per purpose

In [ ]:
uber_drives.groupby(['PURPOSE*']).sum()

new_dataset = {'Airport/Travel':16.5, 'Between Offices':197.0, 'Charity($)':15.1, 'Commute':180.2, 'Customer Visit':2089.5,
              'Errand/Supplies':508.0, 'Meal/Entertain':911.7, 'Meeting':2851.3, 'Moving':18.2, 'Temporary Site':523.7}


purpose = list(new_dataset.keys())
miles = list(new_dataset.values())

fig = plt.figure(figsize = (17, 10))

plt.bar(purpose, miles, color ='green', width = 0.4)
plt.title("Total/Sum miles per purpose")
plt.xlabel("purpose")
plt.ylabel("miles")

plt.show()

## Q20. Display a dataframe of Purpose and the distance travelled for that particular Purpose. (3 points)
#### Note: Use the original dataframe without dropping "NA" values

In [ ]:
# Displaying a dataframe of Purpose and the distance travelled for that particular Purpose

uber_drives.groupby(['PURPOSE*']).sum()

## Q21. Plot number of trips vs Category of trips. (3 points)
#### Note: Use the original dataframe without dropping the 'NA' values.
#### Hint : You can make a countplot or barplot.

In [ ]:
uber_drives.groupby(['CATEGORY*']).groups



## Q22. What is proportion of miles that are covered as Business trips and what is the proportion of miles that are covered as Personal trips? (3 points)

### Note:Use the original dataframe without dropping the 'NA' values. The proportion calculation  is with respect to the 'miles' variable.

#### Hint: Proportion of miles covered as business trips= (Total Miles clocked as Business Trips)/ (Total Miles)
#### Proportion of miles covered as personal trips= (Total Miles clocked as Personal Trips)/ (Total Miles)


In [ ]:
total_miles = uber_drives['MILES*'].sum()    # sum of miles = 12204.7
print('Total Miles = ',total_miles)

print('______________________')

miles_per_category = uber_drives.groupby(['CATEGORY*']).sum()

print('\nMiles per category are shown as below = \n\n', miles_per_category)

print('______________________')

#Proportion of miles covered as business trips= (Total Miles clocked as Business Trips)/ (Total Miles)

Proportion_of_miles_covered_as_business_trips = (11487.0)/(12204.7)
print('Proportion of miles covered as business trips: ',Proportion_of_miles_covered_as_business_trips)

print('______________________')

#Proportion of miles covered as personal trips= (Total Miles clocked as Personal Trips)/ (Total Miles)

Proportion_of_miles_covered_as_personal_trips = (717.7)/(12204.7)
print('Proportion of miles covered as personal trips: ',Proportion_of_miles_covered_as_personal_trips)

